### PR TITLE
v0.4.0 Documentation 

### PR DESCRIPTION
**Pull request by stuartmcalpine**
Contains updates to documentation for branch v0.4.0

Can't currently check the code cells on NERSC until this gets merged to the base branch

### CODE REVIEW

**Comment by JoanneBogart** [comment]
This looks basically ok but since I can't test it I think it's too early for a review.

**Comment by stuartmcalpine** [comment]
The branch is now merged in with the base v0.4.0 branch so the tutorial note books now work (note I had to make my own databae on spin with a v0.4.0 database schema/production schema)

File: docs/source/tutorial_notebooks/getting_started_3_pipelines.ipynb [hunk 1]
```diff
@@ -135,28 +140,28 @@
...
   },
   {
    "cell_type": "markdown",
    "id": "92657fa5",
    "metadata": {},
    "source": [
-    "If you know your executions will only have one dataset associated with them, registering the execution and dataset together during ``register_dataset`` can save a bit of time.\n",
-    "\n",
-    "The ``register_dataset`` function returns both the ``dataset_id`` and associated ``execution_id``."
+    "If you know your executions will only have one dataset associated with them, registering the execution and dataset together during ``Registrar.dataset.register()`` can save a bit of time. You can ofcourse still manually create the execution first, and register the single dataset against it if preferred."
```
Comment (JoanneBogart) [inline docs/source/tutorial_notebooks/getting_started_3_pipelines.ipynb:164]: change "ofcourse" to "of course"

File: docs/source/tutorial_notebooks/getting_started_2_query_datasets.ipynb [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "f48aec2e-2b35-49ed-be76-8818d9e79b2c",
+   "metadata": {},
+   "source": [
+    "## 1) Querying the data registry with a single filter\n",
+    "\n",
+    "Now that we have covered the basics of dataset registration, we can have a look at how to query entries in the database. Note you can only query for datasets within the schema you have connected to.\n",
+    "\n",
+    "We learned how to connect to the DESC data registry in the last tutorial using the `DataRegistry` class, lets connect again using the defaults:"
```
Comment (JoanneBogart) [inline docs/source/tutorial_notebooks/getting_started_2_query_datasets.ipynb:52]: "lets" should be "let's"

File: docs/source/tutorial_notebooks/getting_started_2_query_datasets.ipynb [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+   "source": [
+    "### Query return formats\n",
+    "\n",
+    "Note that three return formats are supported, selected via the optional `return_format` attribute passed to the `find_datasets` function:\n",
+    "\n",
+    "- `return_format=\"property_dict\"` : a dictionary with keys in the format `<table_name>.<column_name>` (default)\n",
+    "- `return_format=\"dataframe\"` : a pandas DataFrame with keys in the format `<table_name>.<column_name>`\n",
+    "- `return_format=\"cursorresult\"` : a SQLAlchemy CursorResult object (see [here](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult) for details)\n",
+    "\n",
+    "Note that for the `CursorResult` object, the property names are still in the format `<table_name>.<column_name>`. Because there is a `.` in the column names, to retrive the properties you need to do `getattr(r, \"dataset.name\")`, where `r` is the row of the `CursorResult` object. "
```
Comment (JoanneBogart) [inline docs/source/tutorial_notebooks/getting_started_2_query_datasets.ipynb:159]: "retrive" should be "retrieve"

File: docs/source/tutorial_notebooks/getting_started_2_query_datasets.ipynb [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "84b25f55-eef4-43b6-9c60-8ada92488dd6",
+   "metadata": {},
+   "source": [
+    "## 2) Querying the data registry with multiple filters\n",
+    "\n",
+    "We are not limited to using a single filter during queries.\n",
+    "\n",
+    "Now lets say we want to return all datasets in the registry with a particular `owner`, that were registered after a certain date. We also want the results in a Pandas dataframe format.\n",
```
Comment (JoanneBogart) [inline docs/source/tutorial_notebooks/getting_started_2_query_datasets.ipynb:189]: "lets" should be "let's" (it's a contraction of "let us")

File: docs/source/tutorial_notebooks/getting_started_1_register_datasets.ipynb [hunk 5]
```diff
@@ -73,11 +70,26 @@
...
+   "cell_type": "markdown",
+   "id": "aa074027-5c6c-48d1-b79f-a6724a1bb386",
+   "metadata": {},
+   "source": [
+    "With no arguments, the `DataRegistry` class will automatically attempt to:\n",
+    "- establish a connection to the registry database using the information in your `~/.config_reg_access` and `~/.pgpass` files\n",
+    "- connect to the default database schema\n",
+    "- use the default NERSC \"`site`\" for the `root_dir`\n",
+    "\n",
+    "The root directory (`root_dir`) is the base path where all ingested data will be copied to. Other than for testing, this should generally be the NERSC `site` address."
```
Comment (JoanneBogart) [inline docs/source/tutorial_notebooks/getting_started_1_register_datasets.ipynb:83]: Maybe a bit clearer to say
"...is the base path under which all ingested data will be copied"
It's the "under which" that matters, to make it clear that we're not overwriting base path.

File: docs/source/tutorial_notebooks/getting_started_1_register_datasets.ipynb [hunk 6]
```diff
@@ -184,45 +204,45 @@
...
+    "By default, datasets in the data registry, once registered, are not overwritable. You can change this behavior by setting `is_overwritable=True` when registering your datasets. If `is_overwritable=True` on one of your previous datasets, you can register a new dataset with the same combination of `relative_path`, `owner` and `owner_type` as before (be warned that any previous data stored under this path will be deleted first). \n",
     "\n",
-    "Note that whilst the data in the shared space will be overwritten with each registration when `is_overwritable=true`, the original entries in the data registry database are never lost (a new unique entry is created each time, and the 'old' entries will obtain `true` for their `is_overwritten` row).\n",
+    "Note that whilst the data in the shared space will be overwritten with each registration when `is_overwritable=True`, the original entries in the data registry database are never lost (a new unique entry is created each time, and the 'old' entries will obtain `True` for their `is_overwritten` row).\n",
```
Comment (JoanneBogart) [inline docs/source/tutorial_notebooks/getting_started_1_register_datasets.ipynb:213]: Change "row" to "column".   That is,
..'old' entries will obtain `True` for their `is_overwritten` **column**

File: docs/source/tutorial_notebooks/production_schema.ipynb [hunk 7]
```diff
@@ -91,13 +93,20 @@
    },
    "outputs": [],
    "source": [
+    "# Production datasets can't be overwritten, so let's generate a random unique name\n",
```
Comment (JoanneBogart) [inline docs/source/tutorial_notebooks/production_schema.ipynb:96]: This is understandable in this context, but normally - and, I would think, especially for production datasets - the name should be meaningful. People may want to use it as a filter when querying.   I'd add a sentence making it clear that normally this is **not** how you would name a production dataset.

**Review by JoanneBogart — APPROVED** [review]
I've left a few comments, mostly addressing typos or similarly minor issues. You should look them over and make adjustments, at least in the couple cases where what's there now is, to my mind, a bit misleading, but I don't think another review is necessary.
The more important thing to address is that, in the first tutorial, it seems that the values for dataset id and execution id returned after overwriting an entry apparently are the ones belonging to the original entry, not the update.   But I think this is a probably with the underlying code, not the notebook.

**Pull request merged**